syntax = "proto3";

package stats;

service StatsService {
  rpc GetPostStats(PostIdRequest) returns (PostStatSummary);

  rpc GetPostViewsByDay(PostIdRequest) returns (DailyStatResponse);

  rpc GetPostLikesByDay(PostIdRequest) returns (DailyStatResponse);

  rpc GetPostCommentsByDay(PostIdRequest) returns (DailyStatResponse);

  rpc GetTopPostsBy(TopRequest) returns (TopItemsResponse);

  rpc GetTopUsersBy(TopRequest) returns (TopItemsResponse);
}

// ----------- Запросы -----------

enum Metric {
  VIEW = 0;
  LIKE = 1;
  COMMENT = 2;
}

message PostIdRequest {
  int32 post_id = 1;
}

message TopRequest {
  Metric metric = 1;
}

// ----------- Ответы -----------

message PostStatSummary {
  int32 views = 1;
  int32 likes = 2;
  int32 comments = 3;
}

message DailyStat {
  string date = 1;
  int32 count = 2;
}

message DailyStatResponse {
  repeated DailyStat items = 1;
}

message TopItem {
  int32 id = 1;
  int32 count = 2;
}

message TopItemsResponse {
  repeated TopItem items = 1;
}